sc-memory/sc-core/src/sc-store/sc_storage.c

```diff
@@ -669,198 +669,216 @@ sc_result _sc_storage_element_erase(sc_addr addr)
+
+    sc_element * connector = sc_hash_table_get(processed_connectors, p_addr);
+    if (connector == null_ptr)
     {
-      p_addr = GUINT_TO_POINTER(SC_ADDR_LOCAL_TO_INT(connector_addr));
+      result = sc_storage_get_element_by_addr(connector_addr, &connector);
+      if (result != SC_RESULT_OK)
+        break;
 
-      sc_element * connector = sc_hash_table_get(cache_table, p_addr);
-      if (connector == null_ptr)
-      {
-        result = sc_storage_get_element_by_addr(connector_addr, &connector);
-        if (result != SC_RESULT_OK)
-          break;
+      sc_hash_table_insert(processed_connectors, p_addr, connector);
+      _sc_storage_element_erase_with_base_element(
+          ctx,
+          connector_chain_begin_addr,
+          connector_addr,
+          processed_connectors,
+          does_branch_have_emitted_events,
+          elements_that_can_be_erased);
+    }
 
-        sc_hash_table_insert(cache_table, p_addr, connector);
-        sc_queue_push(&iter_queue, p_addr);
-      }
+    connector_addr = connector->arc.next_begin_out_arc;
+  }
 
-      connector_addr = connector->arc.next_end_in_arc;
+  connector_addr = el->first_in_arc;
+  while (SC_ADDR_IS_NOT_EMPTY(connector_addr))
+  {
+    sc_pointer p_addr = GUINT_TO_POINTER(SC_ADDR_LOCAL_TO_INT(connector_addr));
+
+    sc_element * connector = sc_hash_table_get(processed_connectors, p_addr);
+    if (connector == null_ptr)
+    {
+      result = sc_storage_get_element_by_addr(connector_addr, &connector);
+      if (result != SC_RESULT_OK)
+        break;
+
+      sc_hash_table_insert(processed_connectors, p_addr, connector);
+      _sc_storage_element_erase_with_base_element(
+          ctx,
+          connector_chain_begin_addr,
+          connector_addr,
+          processed_connectors,
+          does_branch_have_emitted_events,
+          elements_that_can_be_erased);
     }
 
-    sc_monitor_release_read(monitor);
+    connector_addr = connector->arc.next_end_in_arc;
   }
 
-  sc_queue_destroy(&iter_queue);
-  sc_hash_table_destroy(cache_table);
+  sc_monitor_release_write(monitor);
 
-  while (!sc_queue_empty(&addrs_with_not_emitted_erase_events))
-  {
-    sc_addr_hash addr_int = (sc_pointer_to_sc_addr_hash)sc_queue_pop(&addrs_with_not_emitted_erase_events);
-    addr.seg = SC_ADDR_LOCAL_SEG_FROM_INT(addr_int);
-    addr.offset = SC_ADDR_LOCAL_OFFSET_FROM_INT(addr_int);
+  if (!*does_branch_have_emitted_events)
+    sc_list_push_back(elements_that_can_be_erased, (sc_addr_hash_to_sc_pointer)SC_ADDR_LOCAL_TO_INT(addr));
+  return SC_RESULT_OK;
+}
 
+sc_result sc_storage_element_erase(sc_memory_context const * ctx, sc_addr addr)
+{
+  sc_hash_table * connectors_added_to_queue = sc_hash_table_init(g_direct_hash, g_direct_equal, null_ptr, null_ptr);
+  sc_bool does_branch_have_emitted_events = SC_FALSE;
+  sc_list * elements_that_can_be_erased;
+  sc_list_init(&elements_that_can_be_erased);
+
+  sc_result result = _sc_storage_element_erase_with_base_element(
```

Comment:
We need to think of a name for this method

Reply:
yes, _sc_storage_element_erase_with_incoming_outgoing_connectors?